### ISSUE TITLE
Document default behavior of DicomFile.Open

### ISSUE THREAD

**Issue by jovinson-ms** [fo-dicom/fo-dicom#1295]
The `DicomFile.Open*` methods define a read behavior via the `FileReadOptions` enum. Currently, the first enum value is called `Default`, but the documentation only says "Default behaviour." Of course, someone who is familiar with the development history of `fo-dicom`, or who looks through the code to see what decisions are made according to these values will realize that the default behavior is actually the same as the second enum value: `ReadLargeOnDemand`, according to its usage in [StreamByteSource](https://github.com/fo-dicom/fo-dicom/blob/4eb7674ca4cfff32a29f7c072e567d8f36d6ecd9/FO-DICOM.Core/IO/StreamByteSource.cs#L48) and [FileByteSource](https://github.com/fo-dicom/fo-dicom/blob/4eb7674ca4cfff32a29f7c072e567d8f36d6ecd9/FO-DICOM.Core/IO/FileByteSource.cs#L54).

My proposal: why not remove the `Default` value from the enum, and simply state in the documentation for `ReadLargeOnDemand` that it is the default value? The comments mention the possibility of setting this value via configuration in the future, but for now the `Default` value doesn't serve as a single touch point since there are multiple files swapping `Default` for `ReadLargeOnDemand`. This would touch a small number of files, and improve the developer experience with Intellisense. I'd be happy to submit a PR for this, but wanted to hear your thoughts.

tagging: @gofal

**Comment by gofal** [comment]
Hi, thanks for your proposal.
The `Default` entry in the enum is meant to be the default. A user calling the Open-Method with parameter `FileReadOptions.Default` tells fo-dicom just to use the option where fo-dicom internally considers to be the optimal. But when the option `ReadlargeOnDemand` is called, then the user wants this option explicitly.

From the API-Design point of view those two options are different: Let's say fo-dicom continues to be improved and we implement a new option like `FileReadOptions.NeverLoadToMemory` where fo-dicom always works directly on filestream and never loads the pixels into memory.
If a user opens a file with `FileReadoptions.Default`, then it is up to fo-dicom to use the beste way (ReadlargeOnDemand for files < 1 GB, but neverLoadToMemory for files > 1GB - or whatever strategy we consider optimal). 
But if a user opens a file with `FileReadoptions.ReadlargeOnDemand`, when we always have to use that behavior.

When it comes to passing various options into the API, and this options are considered to grow in the future, then it requires some disctinction between explicitly calling an option or telling fo-dicom to use the current default option which also may change in future.

And you are right, the comments in the code tell that is planned to set the default-behavior of `FileReadoptions` gobally, but currently this is hardcoded in various files. So that's still to do. 

Does that sound reasonable to you? Maybe it's just about making a better documentation like:

        /// <summary>
        ///  Use the current default option or let fo-dicom decide which option to use. 
        /// </summary>
        Default,

**Comment by mrbean-bremen** [comment]
>     ///  Use the current default option or let fo-dicom decide which option to use. 

That should be "and", I think; or maybe something like:
```
 /// Use the current default option as selected by fo-dicom. 
```

**Comment by jovinson-ms** [comment]
Thanks, @gofal for the quick reply. 

I guess my overall point is that we should communicate to the end user what the default behavior is, so if we're keeping the `Default` value, my suggestion would be to change the documentation to reflect its current behavior, like:

```dotnet
    /// <summary>
    ///  Use the current default option: read only small tags, but the stream can be kept open to read large tags on demand.
    /// </summary>
    Default,
```

This documentation can be changed if the default behavior changes, but at any given moment, developers will be able to understand the current behavior without digging through code.

That's a good point about the default behavior being a distinct concept from other approaches, but as a developer I would expect the default behavior of a given exposed method to be implicit, like it is in the `Save*` [methods in the same class](https://github.com/fo-dicom/fo-dicom/blob/11cf37891fe8c54f50f344766a3b0382c549ddae/FO-DICOM.Core/DicomFile.cs#L123). 

In other words, when would I expect a call to `Open()` to behave differently from `Open(FileReadOptions.Default)`?

**Comment by gofal** [comment]
Sure, `Open()` and `Open(FileReadOption.Default)` behave the same way, and noone would call that second version, just like in the same class `Save("filename")` and `Save("filename", null)` behave the same and noone would call the second version.
There has to be some distinction between a user calling one of the options explicitly or just letting the default value. In the `Save*` methods the option is a class because it are several different parameters you can set, and there the default value is simply `null`. In the `Open` method the option is just an enum because there are no more other options, and there the default value is just `0` or in enum `Default`. 

So in your understanding the `null` value is kind of implicit enough for having some kind of default behavior. But actually with enums, where you don't have null, the `Default` entry with value 0 ist just the same.
You'd prefer an api like `Open(string filename, FileReadOption? readOption = null, ...)`. That would also be possible. But you must be aware that `Open` has many method overloads, and the more overloads, the more you have to take care about the types of the parameters. Having an enum as parameter makes it pretty much clear, on which position the argument is. But having a nullable enum where the user wants to have the default behavior, leads to `null` as parameter, where the compiler then may not so easily distinct between parameters etc. But thats just some personal experience.

Wanna say thanks for this input here. I like to discuss about design principles and what various arguments pro and con there are.

**Comment by jovinson-ms** [comment]
@gofal thanks for the clarification - you're right, there are so many tradeoffs to balance in API design. Unfortunately, almost everyone has a different solution!

What do you think about just changing the documentation of the `Default` option to mirror the current behavior, as I suggested above? That would take care of my concern that the default behavior is hard to discover.
